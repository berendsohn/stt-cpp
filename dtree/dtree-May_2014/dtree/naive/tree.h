// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
