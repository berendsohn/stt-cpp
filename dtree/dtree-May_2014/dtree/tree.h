// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
