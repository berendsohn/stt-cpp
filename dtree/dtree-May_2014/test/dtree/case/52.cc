#define DTREE_CONFIG EndTreeWithDesc<WithEvert<WithEvert<WithStaticValue<FreeValue, WithAncDescAggr<dtree::DpMin<double>, dtree::DwLess_<double>, WithAncValue<dtree::DpAdd<double>, Begin > > > > > >
#include "test/dtree/tree.cc"
