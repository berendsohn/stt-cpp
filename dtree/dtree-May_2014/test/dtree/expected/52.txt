<<<
  1: vL-6R-8F0 vfaf,000,770,000 -> 3
  2: vL5R3F1 vac3,000,476,000 -> 3
  3: vL29R-9F1 ve65,000,899,000 -> 5
  4: vL0R-9F1 v939,000,b4b,000
  5: vL14R-18F1 vf59,000,807,000 -> 7
  6: vL-6R0F1 v618,000,bf4,000
  7: vL16R1F1 v17e,000,51c,000
  8: vL1R2F1 vb10,000,206,000
  9: vL2R-2F0 v6e3,000,96e,000
  10: vL7R-3F0 v3e0,000,fe7,000
  11: vL3R5F0 v4a2,000,0ab,000
  12: vL7R7F0 v103,000,d81,000 -> 13
  13: vL-5R-7F0 v2e3,000,1c6,000 -> 3
  14: vL-9R4F1 vaf5,000,260,000
  15: vL6R-8F1 va81,000,834,000
>>>
<<<
  @2: vL5R3F1 vL5R3F1 aL0R0 oLinfRinf vac3,000,476,000 f1 f0
      /7: vL10R-13F0 vL10R-13F0 aL-19R-2 oLinfRinf v17e,000,51c,000 f1 f1
        \5: vL-19R-2F0 vL-19R-2F0 aL0R0 oLinfRinf vf59,000,807,000 f1 f0
    \3: vL-12R24F0 vL-12R24F0 aL-9R-15 oL-6R-8 ve65,000,899,000 f0 f1
     ->13: vL-5R-7F0 vL-5R-7F0 aL-5R-7 oL-6R-8 v2e3,000,1c6,000 f1 f1
        \12: vL12R14F0 vL12R14F0 aL0R0 oLinfRinf v103,000,d81,000 f1 f0
       '.1: vL-6R-8F0 vL-6R-8F0 aL-6R-8 oL-6R-8 vfaf,000,770,000 f1 f1

  @4: vL0R-9F1 vL0R-9F1 aL0R0 oLinfRinf v939,000,b4b,000 f0 f0

  @6: vL-6R0F1 vL-6R0F1 aL0R0 oLinfRinf v618,000,bf4,000 f1 f0

  @8: vL1R2F1 vL1R2F1 aL0R0 oLinfRinf vb10,000,206,000 f0 f0

  @9: vL2R-2F0 vL2R-2F0 aL-5R-1 oL-6R-8 v6e3,000,96e,000 f1 f0

  @10: vL7R-3F0 vL7R-3F0 aL0R0 oLinfRinf v3e0,000,fe7,000 f1 f1

  @11: vL3R5F0 vL3R5F0 aL0R0 oLinfRinf v4a2,000,0ab,000 f1 f0

  @14: vL-9R4F1 vL-9R4F1 aL0R0 oLinfRinf vaf5,000,260,000 f0 f1

  @15: vL6R-8F1 vL6R-8F1 aL0R0 oLinfRinf va81,000,834,000 f0 f0
>>>
